operator/duties/attester.go

```diff
@@ -0,0 +1,262 @@
+			}
+		}
+	}
+}
+
+func (h *AttesterHandler) processFetching(ctx context.Context, logger *zap.Logger, epoch phase0.Epoch, slot phase0.Slot) {
+	ctx, cancel := context.WithDeadline(ctx, h.network.Beacon.GetSlotStartTime(slot+1).Add(100*time.Millisecond))
+	defer cancel()
+
+	if h.fetchCurrentEpoch {
+		if err := h.fetchDuties(ctx, logger, epoch); err != nil {
+			logger.Error("failed to fetch duties for current epoch", zap.Error(err))
+			return
+		}
+		h.fetchCurrentEpoch = false
+	}
+
+	if h.fetchNextEpoch && h.shouldFetchNexEpoch(slot) {
+		if err := h.fetchDuties(ctx, logger, epoch+1); err != nil {
+			logger.Error("failed to fetch duties for next epoch", zap.Error(err))
+			return
+		}
+		h.fetchNextEpoch = false
+	}
+}
+
+func (h *AttesterHandler) processExecution(logger *zap.Logger, epoch phase0.Epoch, slot phase0.Slot) {
+	// range over duties and execute
+	if slotMap, ok := h.duties.m[epoch]; ok {
+		if duties, ok := slotMap[slot]; ok {
+			toExecute := make([]*spectypes.Duty, 0, len(duties)*2)
+			for _, d := range duties {
+				if h.shouldExecute(logger, d) {
+					toExecute = append(toExecute, h.toSpecDuty(d, spectypes.BNRoleAttester))
+					toExecute = append(toExecute, h.toSpecDuty(d, spectypes.BNRoleAggregator))
+				}
+			}
+			h.executeDuties(logger, toExecute)
+		}
+	}
+}
+
+func (h *AttesterHandler) fetchDuties(ctx context.Context, logger *zap.Logger, epoch phase0.Epoch) error {
+	start := time.Now()
+	indices := h.validatorController.ActiveIndices(logger, epoch)
+	duties, err := h.beaconNode.AttesterDuties(ctx, epoch, indices)
+	if err != nil {
+		return fmt.Errorf("failed to fetch attester duties: %w", err)
+	}
+
+	specDuties := make([]*spectypes.Duty, 0, len(duties))
+	for _, d := range duties {
+		h.duties.Add(epoch, d.Slot, d)
+		specDuties = append(specDuties, h.toSpecDuty(d, spectypes.BNRoleAttester))
+	}
+
+	logger.Debug("🗂 got duties",
+		fields.Count(len(duties)),
+		fields.Epoch(epoch),
+		fields.Duties(epoch, specDuties),
+		fields.Duration(start))
+
+	// calculate subscriptions
+	subscriptions := calculateSubscriptionInfo(duties)
+	if len(subscriptions) > 0 {
+		if err := h.beaconNode.SubscribeToCommitteeSubnet(subscriptions); err != nil {
+			logger.Warn("failed to submit beacon committee subscription", zap.Error(err))
+		}
+	}
+
+	return nil
+}
+
+func (h *AttesterHandler) toSpecDuty(duty *eth2apiv1.AttesterDuty, role spectypes.BeaconRole) *spectypes.Duty {
+	return &spectypes.Duty{
+		Type:                    role,
+		PubKey:                  duty.PubKey,
+		Slot:                    duty.Slot,
+		ValidatorIndex:          duty.ValidatorIndex,
+		CommitteeIndex:          duty.CommitteeIndex,
+		CommitteeLength:         duty.CommitteeLength,
+		CommitteesAtSlot:        duty.CommitteesAtSlot,
+		ValidatorCommitteeIndex: duty.ValidatorCommitteeIndex,
+	}
+}
+
+func (h *AttesterHandler) shouldExecute(logger *zap.Logger, duty *eth2apiv1.AttesterDuty) bool {
+	currentSlot := h.network.Beacon.EstimatedCurrentSlot()
```

Comment:
when you call this function you usually already know the current slot, so passing it in instead of calling `h.network.Beacon.EstimatedCurrentSlot()` can reduce times you call it and reduce the dependency in this function.

Reply:
We explicitly estimate the current slot at the exact time of the function call to ensure our operation is based on the most accurate and recent data